### ISSUE TITLE
Case syntax with optional parameter doesn't work on v0.23

### ISSUE THREAD

**Issue by thucnc** [metabase/metabase#4581]
Hello,
I got below error when using 'case' syntax with optional parameter on Postgresql db (it used to work on v0.22)
```
[[and case when {% raw %}{{status}}{% endraw %} ='CANCELLED' then status = 'CANCELLED'
    else status in ('COMPLETED', 'FAILED') end]]
```
![screen shot 2017-03-22 at 3 09 20 pm](https://cloud.githubusercontent.com/assets/2088055/24187889/adaa8046-0f11-11e7-9626-e9a3197be974.png)

The same error when I removed {% raw %}{% endraw %}

**Comment by salsakran** [comment]
Dumb question but what is {% raw %} targeted for? We don't have that in our template language as far as I know.

By my read, this is a case where our closing off an avenue for SQL injection (namely moving from string insertion to SQL parameters) probably broke a class of queries. This class (namely where the actual SQL is being generated, vs representing parameters) is a little tricky and I'm not sure we'll be able to unbreak it without re-introducing an attack vector. 

If you can insert a field name (i.e. a sql token), you can also insert `fieldname=false;drop table users;` . In general, we'd rather have a whitelist of allowable SQL constructs we know are safe, than develop a blacklist of attacks (which we'll forever be behind on).

If you describe what you're trying to do, we can extend our template language to try to accommodate it!

**Comment by camsaul** [comment]
yeah unfortunately since we're not directly substituting the params into the SQL more something like the query above won't work anymore

**Comment by salsakran** [comment]
Also note that because we're now exposing parameters in public dashboards etc, someone without an account on your Metabase instance could be feeding that SQL template parameters as opposed to just you or your users, if you publish it in a blog post or your application.

**Comment by thucnc** [comment]
@salsakran , I found it on your documentation here: https://github.com/metabase/metabase/blob/master/docs/users-guide/12-sql-parameters.md
![screen shot 2017-03-22 at 2 16 51 pm](https://cloud.githubusercontent.com/assets/2088055/24232807/257712fc-0fc0-11e7-80e5-431ad1dbdbc2.png)

On the other hand, if we can't fix this issue, may you invent a if-else syntax to handle such case ? As we badly depend on this feature on our queries to be more flexible and simple.

Thanks.

**Comment by salsakran** [comment]
Ahh, that's markup for our docs site, it is needed to get curly brackets `{{` and `}}` to show up in markdown. 

See http://www.metabase.com/docs/latest/users-guide/12-sql-parameters.html

The github docs directory should be viewed as authoritative for the state of the branch you're on, but not necessarily perfectly formatted for viewing. The metabase.com/docs docs are the "canonical source"

Hmm .... what values are you feeding to "{{status}}" above?

**Comment by thucnc** [comment]
@salsakran thank you for your explanation.

In our broken query, {{status}} belongs to a set of constants such as [FAILED, CANCELLED, COMPLETED], but we want to set a condition that when no value is set to {{status}}, we will filter all statuses having COMPLETED or FAILED. 

So, such scenario needs an if-else condition syntax for {{status}}.

**Comment by salsakran** [comment]
Are end users typing out "failed", "cancelled", etc?

**Comment by thucnc** [comment]
@salsakran yes, they will type such words.

**Comment by salsakran** [comment]
as mentioned in #4545 you can no have `{{status}}` mapped to the underlying column via a field filter clause, and your users will be able to do the same thing they used to. For now, the default value can only be a single value however.

**Comment by thucnc** [comment]
@salsakran, I'm not sure if I explained correctly my desire, so let me rephrase it. My users can enter ONE of these statuses ['CANCELED','COMPLETED',FAILED',...] or they can leave it empty. If {{status}} has value, we just simply set the condition `status={{status}}`, but when the {{status}} is empty, we want the condition as `status in ('COMPLETED','FAILED')`.

So this way make our query is more flexible. May you advise me how to achieve this, even with workaround ?

**Comment by flamber** [comment]
Using `coalesce` and optional syntax, it's possible to do. Example using Sample Dataset on 0.35.3:
```sql
SELECT COUNT(*) FROM PRODUCTS
WHERE CASE WHEN COALESCE([[{{filter}},]] NULL)='Gizmo' THEN CATEGORY='Gizmo' ELSE CATEGORY IN ('Gadget', 'Widget') END
```
